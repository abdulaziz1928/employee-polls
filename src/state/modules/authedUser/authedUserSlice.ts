import { PayloadAction, createSlice } from "@reduxjs/toolkit";

const initialState: string | null = null;

const authedUserSlice = createSlice({
  name: "authedUser",
  initialState: initialState as string | null,
  reducers: {
    setAuthedUser: (_, action: PayloadAction<string>) => action.payload,
    logoutAuthedUser: () => null,
  },
});

export const { setAuthedUser, logoutAuthedUser } = authedUserSlice.actions;

export default authedUserSlice.reducer;
