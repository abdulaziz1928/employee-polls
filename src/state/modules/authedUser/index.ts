import reducer from "./authedUserSlice";
export * from "./authedUserSlice";
export default reducer;
