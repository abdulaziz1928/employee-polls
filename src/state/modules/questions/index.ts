import reducer from "./questionsSlice";
export * from "./questionsSlice";
export * from "./models/question";
export default reducer;
