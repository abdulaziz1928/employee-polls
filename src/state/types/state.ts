import LoadingStatus from "./loading_status";

export default interface IState<T> {
  entities: T;
  loading: LoadingStatus;
  error?: string;
}
