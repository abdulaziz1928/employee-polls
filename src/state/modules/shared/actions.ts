import { Dispatch } from "@reduxjs/toolkit";
import { setAuthedUser } from "../authedUser";
import { getInitialData } from "../../apis/api";
import { receiveQuestions } from "../questions";
import { receiveUsers } from "../users";

const AUTHED_USER_ID = "tylermcginnis";

export function handleInitialData() {
  return (dispatch: Dispatch) => {
    return getInitialData().then(({ users, questions }) => {
      dispatch(receiveUsers(users));
      dispatch(receiveQuestions(questions));
      dispatch(setAuthedUser(AUTHED_USER_ID));
    });
  };
}
