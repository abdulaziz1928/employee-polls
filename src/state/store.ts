import authedUser from "./modules/authedUser";
import questions from "./modules/questions";
import users from "./modules/users";
import logger from "./middlewares/logger";
import prefrences from "./modules/prefrences/prefrencesSlice";
import snackbar from "./modules/snackbar/snackbarSlice";
import {
  PreloadedState,
  combineReducers,
  configureStore,
} from "@reduxjs/toolkit";

const reducers = { authedUser, users, questions, prefrences, snackbar };

const rootReducer = combineReducers(reducers);

const middleWares = [logger];

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    middleware: (gDM) => gDM().concat(...middleWares),
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore["dispatch"];
