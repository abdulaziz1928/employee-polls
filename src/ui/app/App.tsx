import { useEffect } from "react";
import { useAppDispatch } from "./hooks";
import { handleInitialData } from "../../state/modules/shared/actions";
import { CssBaseline } from "@mui/material";
import { Routes, Route, Navigate } from "react-router-dom";
import PageRoutes from "../../state/types/page_routes";
import HomePage from "../pages/home_page";
import LeaderboardPage from "../pages/leaderboard_page";
import NewPollPage from "../pages/new_poll_page";
import PollPage from "../pages/poll_page";
import Signin from "../pages/login_page";
import PrivateRoutes from "../components/common/private_routes";
import SuccessSnackbar from "../components/common/success_snackbar";
import NotFoundPage from "../pages/not_found_page";

export default function App() {
  const dispatch = useAppDispatch();
  useEffect(() => {
    dispatch(handleInitialData());
  }, [dispatch]);
  return (
    <>
      <CssBaseline />
      <SuccessSnackbar />
      <Routes>
        <Route element={<PrivateRoutes />}>
          <Route path={PageRoutes.Home} element={<HomePage />} />
          <Route path={PageRoutes.Leaderboard} element={<LeaderboardPage />} />
          <Route path={PageRoutes.Question} element={<PollPage />} />
          <Route path={PageRoutes.New} element={<NewPollPage />} />
          <Route path={PageRoutes.NotFound} element={<NotFoundPage />} />
          <Route path="*" element={<Navigate to={PageRoutes.NotFound} />} />
        </Route>
        <Route path={PageRoutes.Login} element={<Signin />} />
      </Routes>
    </>
  );
}
