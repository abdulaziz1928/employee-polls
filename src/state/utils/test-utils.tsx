import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { setupStore, AppStore, RootState } from "../store";
import { MemoryRouter } from "react-router-dom";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithProvidersAndRouter(
  ui: React.ReactElement,

  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route: string = "/"
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <MemoryRouter initialEntries={[route]}>{children}</MemoryRouter>
      </Provider>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
