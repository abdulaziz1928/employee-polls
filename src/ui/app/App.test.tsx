import { screen } from "@testing-library/react";
import App from "./App";
import { renderWithProvidersAndRouter } from "../../state/utils/test-utils";
import { AppStore, setupStore } from "../../state/store";
import { handleInitialData } from "../../state/modules/shared/actions";
import { setAuthedUser } from "../../state/modules/authedUser";

let store: AppStore;

describe("App", () => {
  beforeAll(async () => {
    store = setupStore();
    await store.dispatch(handleInitialData());
  });

  it("should render the component", () => {
    const { container } = renderWithProvidersAndRouter(<App />);
    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it("should render the login page if user is not logged in", () => {
    renderWithProvidersAndRouter(<App />);
    expect(screen.getByTestId("login-container")).toBeInTheDocument();
    expect(screen.queryByTestId("header-appbar")).not.toBeInTheDocument();
  });

  it("should render the home page if user is logged in", async () => {
    store.dispatch(setAuthedUser("tylermcginnis"));
    renderWithProvidersAndRouter(<App />, { store });
    expect(screen.getByTestId("home-container")).toBeInTheDocument();
  });

  it("should render the header if user is logged in", async () => {
    renderWithProvidersAndRouter(<App />, { store });
    expect(screen.getByTestId("header-appbar")).toBeInTheDocument();
  });
});

