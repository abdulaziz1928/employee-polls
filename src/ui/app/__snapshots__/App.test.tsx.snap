// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render the component 1`] = `
<div>
  <span
    class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate css-eglki6-MuiLinearProgress-root"
    role="progressbar"
  >
    <span
      class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate css-5ir5xx-MuiLinearProgress-bar1"
    />
    <span
      class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate css-1r8wrcl-MuiLinearProgress-bar2"
    />
  </span>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiContainer-root MuiContainer-maxWidthSm css-wg44nm-MuiPaper-root-MuiContainer-root"
    data-testid="login-container"
  >
    <div
      class="MuiBox-root css-0"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-alignCenter css-47suzj-MuiTypography-root"
      >
        Sign in
      </h1>
    </div>
    <form
      class="MuiBox-root css-0"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <div
          class="MuiStack-root css-1ucnu3w-MuiStack-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              id="username"
            >
              Username
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl  css-zethra-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              required=""
            >
              <div
                aria-disabled="true"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="username mui-component-select-username"
                class="MuiSelect-select MuiSelect-outlined Mui-disabled MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="mui-component-select-username"
                role="button"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                disabled=""
                name="username"
                required=""
                tabindex="-1"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined Mui-disabled css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    username
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-186q1z4-MuiTypography-root"
            >
              This is a demo app so feel free to choose one of the predefined user accounts
            </p>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Sign in
            </h6>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
